### PR TITLE
Remove ImmutableOpenMap from snapshot services

### PR DESCRIPTION
**Pull request by thecoop**
Part of #86239

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/snapshots/InternalSnapshotsInfoService.java [hunk 1]
```diff
@@ -64,7 +65,7 @@ public class InternalSnapshotsInfoService implements ClusterStateListener, Snaps
     private final Supplier<RerouteService> rerouteService;
 
     /** contains the snapshot shards for which the size is known **/
-    private volatile ImmutableOpenMap<SnapshotShard, Long> knownSnapshotShards;
+    private volatile Map<SnapshotShard, Long> knownSnapshotShards;
```
Comment (thecoop) [inline server/src/main/java/org/elasticsearch/snapshots/InternalSnapshotsInfoService.java:68]: Is performance an important factor in this class?

File: server/src/main/java/org/elasticsearch/snapshots/InternalSnapshotsInfoService.java [hunk 2]
```diff
@@ -112,14 +113,15 @@ private void setMaxConcurrentFetches(Integer maxConcurrentFetches) {
...
+            Map<SnapshotShard, Long> snapshotShardSizes = knownSnapshotShards;
             if (failedSnapshotShards.isEmpty() == false) {
+                snapshotShardSizes = new HashMap<>(snapshotShardSizes); // create a new map to modify
                 for (SnapshotShard snapshotShard : failedSnapshotShards) {
                     Long previous = snapshotShardSizes.put(snapshotShard, ShardRouting.UNAVAILABLE_EXPECTED_SHARD_SIZE);
                     assert previous == null : "snapshot shard size already known for " + snapshotShard;
                 }
             }
-            return new SnapshotShardSizeInfo(snapshotShardSizes.build());
+            return new SnapshotShardSizeInfo(snapshotShardSizes);
```
Comment (grcevski) [inline server/src/main/java/org/elasticsearch/snapshots/InternalSnapshotsInfoService.java:124]: Should we be passing in here Collections.unmodifiableMap(snapshotShardSizes) instead of the mutable HashMap?

File: server/src/main/java/org/elasticsearch/snapshots/InternalSnapshotsInfoService.java [hunk 3]
```diff
@@ -226,12 +228,10 @@ protected void doRun() throws Exception {
...
                 assert removed : "snapshot shard to remove does not exist " + snapshotShardSize;
                 if (isMaster) {
-                    final ImmutableOpenMap.Builder<SnapshotShard, Long> newSnapshotShardSizes = ImmutableOpenMap.builder(
-                        knownSnapshotShards
-                    );
+                    final Map<SnapshotShard, Long> newSnapshotShardSizes = new HashMap<>(knownSnapshotShards);
                     updated = newSnapshotShardSizes.put(snapshotShard, snapshotShardSize) == null;
                     assert updated : "snapshot shard size already exists for " + snapshotShard;
-                    knownSnapshotShards = newSnapshotShardSizes.build();
+                    knownSnapshotShards = Map.copyOf(newSnapshotShardSizes);
```
Comment (grcevski) [inline server/src/main/java/org/elasticsearch/snapshots/InternalSnapshotsInfoService.java:234]: I believe Collections.unmodifiableMap might be a better choice here, since we don't modify newSnapshotShardSizes after. We shouldn't need to make a copy I think.

File: server/src/main/java/org/elasticsearch/snapshots/InternalSnapshotsInfoService.java [hunk 4]
```diff
@@ -269,17 +269,17 @@ public void onAfter() {
...
                 if (newSnapshotShardSizes == null) {
-                    newSnapshotShardSizes = ImmutableOpenMap.builder(knownSnapshotShards);
+                    newSnapshotShardSizes = new HashMap<>(knownSnapshotShards);
                 }
                 newSnapshotShardSizes.remove(shard);
             }
         }
         if (newSnapshotShardSizes != null) {
-            knownSnapshotShards = newSnapshotShardSizes.build();
+            knownSnapshotShards = Map.copyOf(newSnapshotShardSizes);
```
Comment (grcevski) [inline server/src/main/java/org/elasticsearch/snapshots/InternalSnapshotsInfoService.java:282]: I think we can also use Collections.unmodifiableMap here.

File: server/src/main/java/org/elasticsearch/snapshots/RestoreService.java [hunk 5]
```diff
@@ -723,19 +722,18 @@ public static RestoreInProgress updateRestoreStateWithDeletedIndices(RestoreInPr
...
+                    if (shards == null) {
+                        shards = new HashMap<>(entry.shards());
                     }
-                    shardsBuilder.put(shardId, new ShardRestoreStatus(null, RestoreInProgress.State.FAILURE, "index was deleted"));
+                    shards.put(shardId, new ShardRestoreStatus(null, RestoreInProgress.State.FAILURE, "index was deleted"));
                 }
             }
-            if (shardsBuilder != null) {
-                ImmutableOpenMap<ShardId, ShardRestoreStatus> shards = shardsBuilder.build();
+            if (shards != null) {
```
Comment (grcevski) [inline server/src/main/java/org/elasticsearch/snapshots/RestoreService.java:736]: I think we need to wrap the mutable HashMap in Collections.unmodifiableMap if it's not null down below when we create a new RestoreInProgress.Entry.

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java [hunk 6]
```diff
@@ -673,7 +672,7 @@ public ClusterState execute(ClusterState currentState) {
                                 }
                             }
                         }
-                        updatedEntry = cloneEntry.withClones(clonesBuilder.build());
+                        updatedEntry = cloneEntry.withClones(clones);
```
Comment (grcevski) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:675]: Wrap in Collections.unmodifiableMap.

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java [hunk 7]
```diff
@@ -1173,13 +1172,13 @@ public ClusterState execute(ClusterState currentState) {
                                     }
                                     if (clones != null) {
                                         changed = true;
-                                        updatedEntriesForRepo.add(snapshot.withClones(clones.build()));
+                                        updatedEntriesForRepo.add(snapshot.withClones(clones));
```
Comment (grcevski) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:1175]: Another wrap.

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java [hunk 8]
```diff
@@ -1356,7 +1355,7 @@ private static ImmutableOpenMap<ShardId, ShardSnapshotStatus> processWaitingShar
             }
         }
         if (snapshotChanged) {
-            return shards.build();
+            return shards;
```
Comment (grcevski) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:1358]: We need to wrap into unmodifiable collection here.

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java [hunk 9]
```diff
@@ -1910,36 +1909,34 @@ public static ClusterState stateWithoutSnapshot(ClusterState state, Snapshot sna
...
     ) {
         if (updatedShardAssignments == null) {
             entries.add(entryToUpdate);
         } else {
-            final ImmutableOpenMap.Builder<ShardId, ShardSnapshotStatus> updatedStatus = ImmutableOpenMap.builder(entryToUpdate.shards());
-            updatedStatus.putAllFromMap(updatedShardAssignments.build());
-            entries.add(entryToUpdate.withShardStates(updatedStatus.build()));
+            final Map<ShardId, ShardSnapshotStatus> updatedStatus = new HashMap<>(entryToUpdate.shards());
+            updatedStatus.putAll(updatedShardAssignments);
+            entries.add(entryToUpdate.withShardStates(updatedStatus));
```
Comment (grcevski) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:1919]: wrap here again with unmodifiable map.

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java [hunk 10]
```diff
@@ -1910,36 +1909,34 @@ public static ClusterState stateWithoutSnapshot(ClusterState state, Snapshot sna
...
             entries.add(entryToUpdate);
         } else {
-            final ImmutableOpenMap.Builder<RepositoryShardId, ShardSnapshotStatus> updatedStatus = ImmutableOpenMap.builder(
-                entryToUpdate.shardsByRepoShardId()
-            );
-            updatedStatus.putAllFromMap(updatedShardAssignments.build());
-            entries.add(entryToUpdate.withClones(updatedStatus.build()));
+            final Map<RepositoryShardId, ShardSnapshotStatus> updatedStatus = new HashMap<>(entryToUpdate.shardsByRepoShardId());
+            updatedStatus.putAll(updatedShardAssignments);
+            entries.add(entryToUpdate.withClones(updatedStatus));
```
Comment (grcevski) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:1933]: Another wrap.

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java [hunk 11]
```diff
@@ -2711,7 +2709,7 @@ private SnapshotsInProgress updatedSnapshotsInProgress(ClusterState currentState
                                     );
                                 }
                             }
-                            snapshotEntries.add(entry.withClones(updatedAssignmentsBuilder.build()));
+                            snapshotEntries.add(entry.withClones(updatedAssignments));
```
Comment (grcevski) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:2712]: Another wrap?

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java [hunk 12]
```diff
@@ -2740,24 +2738,23 @@ private SnapshotsInProgress updatedSnapshotsInProgress(ClusterState currentState
...
                                 } else {
                                     if (updated.isActive()) {
                                         markShardReassigned(shardId, reassignedShardIds);
                                     }
-                                    updatedAssignmentsBuilder.put(sid, updated);
+                                    updatedAssignments.put(sid, updated);
                                 }
                             }
-                            final SnapshotsInProgress.Entry updatedEntry = entry.withShardStates(updatedAssignmentsBuilder.build());
+                            final SnapshotsInProgress.Entry updatedEntry = entry.withShardStates(updatedAssignments);
```
Comment (grcevski) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:2757]: This should we wrapped in unmodifiable collection too.

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java [hunk 13]
```diff
@@ -2777,9 +2774,7 @@ private SnapshotsInProgress updatedSnapshotsInProgress(ClusterState currentState
                 for (IndexId indexIdToRefresh : newIndexIdsToRefresh) {
                     updatedIndexIds.put(indexIdToRefresh, new IndexId(indexIdToRefresh.getName(), UUIDs.randomBase64UUID()));
                 }
-                for (int i = 0; i < snapshotEntries.size(); i++) {
-                    snapshotEntries.set(i, snapshotEntries.get(i).withUpdatedIndexIds(updatedIndexIds));
-                }
+                snapshotEntries.replaceAll(e -> e.withUpdatedIndexIds(updatedIndexIds));
```
Comment (grcevski) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:2777]: Depending on the size of the entries, running this with a lambda might be quite a bit slower.

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java [hunk 14]
```diff
@@ -2897,12 +2892,12 @@ private static ImmutableOpenMap<ShardId, SnapshotsInProgress.ShardSnapshotStatus
...
                         shardSnapshotStatus = initShardSnapshotStatus(shardRepoGeneration, indexRoutingTable.shard(i).primaryShard());
                     }
-                    builder.put(shardId, shardSnapshotStatus);
+                    shards.put(shardId, shardSnapshotStatus);
                 }
             }
         }
 
-        return builder.build();
+        return shards;
```
Comment (grcevski) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:2900]: UnmodifiableMap here too.

File: server/src/main/java/org/elasticsearch/snapshots/InternalSnapshotsInfoService.java (hunk 2 above)
Comment (thecoop) [reply server/src/main/java/org/elasticsearch/snapshots/InternalSnapshotsInfoService.java:124]: Probably - but I'm not sure if the immutable-ness should be handled by the `SnapshotShardSizeInfo` constructor, as it's a public method/class (cf `SnapshotsInProgress.Entry`). How public is this class?

File: server/src/main/java/org/elasticsearch/snapshots/RestoreService.java (hunk 5 above)
Comment (thecoop) [reply server/src/main/java/org/elasticsearch/snapshots/RestoreService.java:736]: `Entry` is a public class, so the constructor should be extra-careful about immutableness in the constructor, just wrapping `Collections.unmodifiableMap` leaves the possibility for the caller to modify the passed-in collection afterwards

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java (hunk 6 above)
Comment (thecoop) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:675]: The `Entry` constructor calls `Map.copyOf` on this map anyway

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java (hunk 7 above)
Comment (thecoop) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:1175]: See above

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java (hunk 8 above)
Comment (thecoop) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:1358]: Possibly...this is a private method, and the map only goes to the `SnapshotsInProgress.Entry` class. However, that constructor doesn't call `Map.copyOf` on thie collection, whereas it does on other arguments. Looks like these classes needs a lot more consistency in its use of maps & immutability ownership now. What are your thoughts?

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java (hunk 13 above)
Comment (thecoop) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:2777]: Possibly, although `ArrayList.replaceAll` has got an implementation that uses indexes rather than a list iterator. Then the only additional cost is calling the lambda - which I expect the JVM would be able to fully inline anyway, as the lambda closure object is not escaping outside this method.

If this is performance sensitive I can revert this, but then should probably put a comment saying this needs to be kept as-is.

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java (hunk 8 above)
Comment (grcevski) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:1358]: Yeah I think we need to ensure there's consistency on how we handle the mutable maps on all of these touch points. The classes that used to receive ImmutableOpenMap before should get an immutable map always, or a copy if needs to happen. As you mentioned, it might be better to do this in the constructors if it's not done already.

We need to prevent hard to track bug where a map is modified after it has been given to another consumer, so perhaps a pattern to follow is something like:

```
var map = new HashMap<...>();

... do something with the map

map = Collections.unmodifiableMap(map) // where we would've normally done ImmutableOpenMap.build().

```

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java (hunk 13 above)
Comment (grcevski) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:2777]: I think it wouldn't be too hard to write a small JMH that will verify if the lambda is slower. I think your version of the code is much nicer, but I'm suspicious of why the code was written as before.

Comment (thecoop) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:2774]: I'll see if I can do a quick check

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java (hunk 8 above)
Comment (thecoop) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java]: I've had a general look at maps in this PR:
- Maps in fields are immutable through `Map.copyOf`
- Maps returned from methods are shallow immutable through `unmodifiableMap`
- Maps that are passed into a method & then go out of scope after don't matter - its up to the method called to copy appropriately if it needs to

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java (hunk 13 above)
Comment (thecoop) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:2775]: `replaceAll` is actually quicker:
```
Benchmark.testIterator    thrpt    5  375361.849 ±  7966.389  ops/s
Benchmark.testReplaceAll  thrpt    5  443378.984 ± 14116.645  ops/s
```

File: server/src/main/java/org/elasticsearch/cluster/RestoreInProgress.java [hunk 15]
```diff
@@ -132,16 +132,12 @@ public Entry(
             List<String> indices,
             Map<ShardId, ShardRestoreStatus> shards
         ) {
+            this.uuid = Objects.requireNonNull(uuid);
             this.snapshot = Objects.requireNonNull(snapshot);
             this.state = Objects.requireNonNull(state);
-            this.quiet = Objects.requireNonNull(quiet);
```
Comment (grcevski) [inline server/src/main/java/org/elasticsearch/cluster/RestoreInProgress.java:137]: Why did we decide to remove this requireNonNull?

File: server/src/main/java/org/elasticsearch/snapshots/InternalSnapshotsInfoService.java [hunk 16]
```diff
@@ -65,9 +66,11 @@ public class InternalSnapshotsInfoService implements ClusterStateListener, Snaps
     private final Supplier<RerouteService> rerouteService;
 
     /** contains the snapshot shards for which the size is known **/
+    // volatile for the unlocked access in numberOfKnownSnapshotShardSizes()
+    // map itself is immutable
     private volatile Map<SnapshotShard, Long> knownSnapshotShards;
 
-    private volatile boolean isMaster;
```
Comment (grcevski) [inline server/src/main/java/org/elasticsearch/snapshots/InternalSnapshotsInfoService.java:69]: While I think this is correct, the boolean is only modified inside a sync block, this change is outside the scope of the PR. We should make a separate PR to address different issues and tag the appropriate folks to review.

File: server/src/main/java/org/elasticsearch/snapshots/InternalSnapshotsInfoService.java [hunk 17]
```diff
@@ -95,7 +98,7 @@ public InternalSnapshotsInfoService(
         this.knownSnapshotShards = Map.of();
         this.unknownSnapshotShards = new LinkedHashSet<>();
         this.failedSnapshotShards = new LinkedHashSet<>();
-        this.queue = new LinkedList<>();
```
Comment (grcevski) [inline server/src/main/java/org/elasticsearch/snapshots/InternalSnapshotsInfoService.java:97]: The array deque is backed by an array, while a LinkedList is chained elements, perhaps there was a reason for this, but either way it should be addressed in a separate PR if you want to make this change.

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java (hunk 13 above)
Comment (grcevski) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:2775]: Great! TIL.

**Review by grcevski — COMMENTED** [review]
Thanks for iterating on this Simon. Can we please separate the improvements to InternalSnapshotInfoService into a separate PR and keep this PR focused on ImmutableOpenMap?

File: server/src/main/java/org/elasticsearch/cluster/RestoreInProgress.java (hunk 15 above)
Comment (thecoop) [reply server/src/main/java/org/elasticsearch/cluster/RestoreInProgress.java:137]: It's called on a `boolean`, which can never be null

File: server/src/main/java/org/elasticsearch/snapshots/InternalSnapshotsInfoService.java (hunk 17 above)
Comment (thecoop) [reply server/src/main/java/org/elasticsearch/snapshots/InternalSnapshotsInfoService.java:97]: Yup, good idea

File: server/src/main/java/org/elasticsearch/snapshots/InternalSnapshotsInfoService.java (hunk 16 above)
Comment (thecoop) [reply server/src/main/java/org/elasticsearch/snapshots/InternalSnapshotsInfoService.java:69]: #90087 for after this PR is merged

**Review by grcevski — APPROVED** [review]
LGTM! I left few comments that I think we should address.
